src/navier.f90

```diff
@@ -1280,5 +1289,271 @@ subroutine tbl_flrt (ux1,uy1,uz1)
+        enddo
+    endif
+
+  end subroutine pipe_bulk
+  !********************************************************************
+  !
+  subroutine pipe_bulk_u(ux,uy,uz,ep,ub_constant)
+  !
+  !********************************************************************
+
+    use decomp_2d
+    use variables
+    use param
+    use var
+    use ibm_param, only: rai
+    use MPI
+
+    implicit none
+    !INPUTS
+    real(mytype),intent(inout),dimension(xsize(1),xsize(2),xsize(3))    :: ux,uy,uz
+    real(mytype),intent(in   ),dimension(xsize(1),xsize(2),xsize(3))    :: ep
+    real(mytype),intent(in   )                                          :: ub_constant !bulk velocity value
+    !LOCALS
+    real(mytype)                                                        :: qm,qmm      !flow rate
+    real(mytype)                                                        :: ym,zm,yc,zc,r
+    real(mytype)                                                        :: ncount,ncountt
+    integer                                                             :: j,i,k,code
+    character(len=30)                                                   :: filename
+    integer, save                                                       :: local_io_unit=-1
+
+    yc = yly/two
+    zc = zlz/two
+
+    if (local_io_unit.eq.-1 .and. nrank.eq.0) then
+       open(newunit=local_io_unit,file='Ub.dat',status='unknown')
+    endif
+
+    !--------------------------- Bulk Velocity ---------------------------
+    !Calculate loss of streamwise mean pressure gradient
+    qm=zero
+    ncount=zero
+    do k=1,xsize(3)
+        zm=dz*real(xstart(3)-1+k-1,mytype)-zc
+        do j=1,xsize(2)
+            if (istret.eq.0) ym=real(j+xstart(2)-1-1,mytype)*dy-yc
+            if (istret.ne.0) ym=yp(j+xstart(2)-1)-yc
+            r=sqrt(ym*ym+zm*zm)
+            do i=1,xsize(1)
+                if (r.le.rai.and.ep(i,j,k).eq.0) then
+                    qm=qm+ux(i,j,k)
+                    ncount=ncount+one
+                endif
+            enddo
+        enddo
+    enddo
+    call MPI_ALLREDUCE(qm,qmm,1,real_type,MPI_SUM,MPI_COMM_WORLD,code)
+    call MPI_ALLREDUCE(ncount,ncountt,1,real_type,MPI_SUM,MPI_COMM_WORLD,code)
+    qmm=qmm/ncountt
+    if (nrank==0) then
+       if (mod(itime, ilist)==0) print *,'Velocity:'
+       if (mod(itime, ilist)==0) print *,'    Bulk velocity before',qmm
+       write(local_io_unit,*) real((itime-1)*dt,mytype), (ub_constant-qmm) !write pressure drop
+    endif
+
+    !Correction
+    do j=1,xsize(2)
+        if (istret.eq.0) ym=real(j+xstart(2)-1-1,mytype)*dy-yc
+        if (istret.ne.0) ym=yp(j+xstart(2)-1)-yc
+        do k=1,xsize(3)
+            zm=dz*real(xstart(3)-1+k-1,mytype)-zc
+            r=sqrt(ym*ym+zm*zm)
+            do i=1,xsize(1)
+                if (r.le.rai.and.ep(i,j,k).eq.0) then
+                    ux(i,j,k)=ux(i,j,k)+(ub_constant-qmm)
+                endif
+            enddo
+        enddo
+    enddo
+
+    !Check new bulk velocity
```

Comment:
Can we move the solid zone / buffer zone with the correction above ?

Then this part should be computed only when `mod(itime, ilist)==0`

Reply:
It is indeed better, thank you for the suggestion. I've moved the cancellation of the solid/buffer zone for both velocity and temperature fields in 3a4dcb01ad1ab5f012f8c38ebd3c52299cb89955.